jni/src/test/java/com/horizen/sc2scnative/Sc2ScTest.java

```diff
@@ -0,0 +1,290 @@
+                r.nextInt(),
+                r.nextInt(),
+                (byte) r.nextInt(),
+                new CustomFieldElementsConfig[0],
+                new CustomBitvectorElementsConfig[0],
+                r.nextLong(),
+                r.nextLong(),
+                generateRandomBytes(r, 1024),
+                Optional.empty(),
+                generateRandomBytes(r, 2000),
+                Optional.empty()));
+    }
+
+    private byte[] root(CommitmentTree ct) {
+        try (FieldElement root = ct.getCommitment().get()) {
+            return root.serializeFieldElement();
+        }
+    }
+
+    private byte[] appendRandomMsgHash(Random r) {
+        try (FieldElement msgHash = FieldElement.createRandom(r)) {
+            msgTree.append(msgHash);
+            return msgHash.serializeFieldElement();
+        }
+    }
+
+    private MerklePath getMsgPath(int leafIndex) {
+        msgTree.finalizeTreeInPlace();
+        return msgTree.getMerklePath(leafIndex);
+    }
+
+    private byte[] getMsgRoot() {
+        msgTree.finalizeTreeInPlace();
+        try (FieldElement root = msgTree.root()) {
+            return root.serializeFieldElement();
+        }
+    }
+
+    @Test
+    public void happyPath() {
+        Random r = new Random();
+        RandomProofData proofData = generateRandomProof(r, 42);
+
+        assertNotNull("Proof creation must be successful", proofData.proof);
+
+        assertEquals(psType, ProvingSystem.getProofProvingSystemType(proofData.proof));
+
+        assertTrue(Sc2Sc.verifyProof(
+                proofData.nextScTxCommitmentsRoot,
+                proofData.currentScTxCommitmentsRoot,
+                proofData.msgHash,
+                proofData.proof,
+                vkPath));
+    }
+
+    private byte[] changeBytes(byte[] data) {
+        data[0] = (byte) (data[0] + 1);
+        return data;
+    }
+
+    @Test
+    public void shouldNotVerifyInvalidInput() {
+        Random r = new Random();
+        RandomProofData proofData = generateRandomProof(r, 42);
+
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        changeBytes(proofData.nextScTxCommitmentsRoot),
+                        proofData.currentScTxCommitmentsRoot,
+                        proofData.msgHash,
+                        proofData.proof,
+                        vkPath));
+
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        proofData.nextScTxCommitmentsRoot,
+                        changeBytes(proofData.currentScTxCommitmentsRoot),
+                        proofData.msgHash,
+                        proofData.proof,
+                        vkPath));
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        proofData.nextScTxCommitmentsRoot,
+                        proofData.currentScTxCommitmentsRoot,
+                        changeBytes(proofData.msgHash),
+                        proofData.proof,
+                        vkPath));
+    }
+
```

Comment:
can we add also a bunch of negative tests:
- next cert points correctly to previous cert hash, but its epoch is not prevcert.epoch +1 => the verification should faill
- next cert points to a different prev cert hash => the verification should faill
- next cert has correct fields for prev cert and epoch number but comes from a different scid =>  the verification should faill

Reply:
Should be handled in rust side. Integration tests should just check that the glue from java and rust is correct.